keps/sig-scheduling/4671-gang-scheduling/README.md

```diff
@@ -0,0 +1,932 @@
+
+
+### API
+
+The `Workload` type will be defined with the following structure:
+```go
+type Workload struct {
+	metav1.TypeMeta
+	metav1.ObjectMeta
+	Spec WorkloadSpec
+	Status WorkloadStatus
+}
+
+type ReplicaMode string
+
+const (
+    ReplicaModeUnreplicated ReplicaMode = "Unreplicated"
+)
+
+// WorkloadSpec describes a workload in a portable way that scheduler and related
+// tools can understand.  
+type WorkloadSpec struct {
+    // ControllerRef points to the true workload, e.g. Deployment.
+    // It is optional to set and is intended to make this mapping easier for
+    // things like CLI tools.
+    ControllerRef *v1.ObjectReference
+    // PodGroups is a list of groups of pods.
+    // Each group may request gang scheduling.
+    PodGroups []PodGroup 
+}
+
+type GangMode string
+const (
+	// GangModeGang means that all pods in this PodGroup need to be scheduled as one gang.
+	GangModeSingle GangMode = "Single"
+
+	// GangModeOff means that all pods in this PodGroup do not need to be scheduled as a gang.
+	GangModeOff GangMode = "Off"
+
+	// GangModeReplicatedGang means that there is a variable number of identical copies of this PodGroup,
+    //  as specified in Replicas, and each copy needs to be independently gang scheduled.
+	GangModeReplicated GangMode = "Replicated"
+)
+
+// GangSchedulingPolicy holds options that affect how gang scheduling of one PodGroup is handled by the scheduler.
+type GangSchedulingPolicy struct {
+    // TODO: Decide between SchedulingTimeoutSeconds and making this field of type time.Duration.
+	SchedulingTimeoutSeconds *int
+	MinCount *int
+}
+
+// PodGroup is a group of pods that may contain multiple shapes (EqGroups) and may contain
+// multiple dense indexes (RankedGroups) and which can optionally be replicated in a variable
+// number of identical copies.
+//
+// TODO: Decide on the naming: PodGroup vs GangGroup.
+type PodGroup struct {
+    Name *string
+    GangMode *GangMode // default is "Single"
+
+    // Optional when GangMode = "ReplicatedGang".
+    // Forbidden otherwise.
+    Replicas int
+
+    // GangSchedulingPolicy defines the options applying to all pods in this gang.
+    // Forbidden if GangMode is set to "Off".
+    GangSchedulingPolicy GangSchedulingPolicy
+}
+
+
+type WorkloadStatus struct {
+  // Necessary status fields TBD.
+}
+```
+
+The individual `PodGroups` and `PodGroup` replicas are treated as independent gangs. As an example, if one of
+the groups can be scheduled and the other can't be - this is exactly what will happen. If the underlying
+user intention was to have either both of them or none of them running, they should form a single group and
+not be split into two. A `LeaderWorkerSet` is a good example of it, where a single `PodGroup` replica consists
+of a single leader and `N` workers and that forms a scheduling (and runtime unit), but workload as a whole
+may consist of a number of such replicas.
+
+### Scheduler Changes
```

Comment:
I'd add a dedicated section for Admission Semantics that elaborate on how the gang scheduling will change the addmision flow. (reservasion, rollback, timeout, retry).